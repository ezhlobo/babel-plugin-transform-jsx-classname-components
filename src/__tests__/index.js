const babel = require('babel-core')
const common = require('common-prefix')
const plugin = require('..').default

const sanitizeSource = (src) => {
  const lines = src.trimRight().split('\n')

  if (lines.length > 1) {
    const minIndent = common(lines.filter(Boolean))

    return src.split('\n').map(line => line.substr(minIndent.length)).join('\n')
  }

  return src
}

const transform = (src = '', options = {}) => babel
  .transform(sanitizeSource(src), {
    ast: false,
    plugins: [
      [plugin, options],
    ],
  })
  .code
  .trim()

it('exists', () => {
  expect(plugin).toBeTruthy()
})

it('does not do anything with the simple code', () => {
  const source = `
    <div>
      <Component className="hello" />
      <Component.Inner className="hello" />
      <Component.Inner className="CamelCasedClassName" />
    </div>
  `

  expect(transform(source)).toMatchSnapshot()
})

it('transforms component if capitalized class name goes first', () => {
  const source = `
    <React.Fragment>
      <Component data-text="extra" className="Inner extra" />
      <Component data-text="extra" className="extra Inner" />
    </React.Fragment>
  `

  expect(transform(source)).toMatchSnapshot()
})

it('transforms computed attributes', () => {
  const source = `
    <React.Fragment>
      <Component className={['Inner', 'simple'].join(' ')} />
      <Component className={['Inner', true && 'simple'].join(' ')} />
    </React.Fragment>
  `

  expect(transform(source)).toMatchSnapshot()
})

it('transforms with closing names', () => {
  const source = `
    <React.Fragment>
      <One>Hello</One>
      <Two className="Inner">Hello</Two>
      <Three>Hello</Three>
      <Four className="Inner"><FourNested>Hello</FourNested></Four>
      <Five><FiveNested className="Inner">Hello</FiveNested></Five>
      <Six className="Inner"><SixNested className="Inner">Hello</SixNested></Six>
    </React.Fragment>
  `

  expect(transform(source)).toMatchSnapshot()
})

it('transforms with unusual names', () => {
  const source = `
    <React.Fragment>
      <A className="B" />
      <A0 className="B0" />
      <CamelCase className="CamelCase" />
      <CamelCase0 className="CamelCase0" />
      <CamelCase className="$CamelCase" />
      <Camel_Case className="_Snake_Case" />
    </React.Fragment>
  `

  expect(transform(source)).toMatchSnapshot()
})

describe('with option "objects"', () => {
  const transformWithObjects = source => transform(source, {
    objects: ['Icons', 'Navigation'],
  })

  it('transforms only specified objects', () => {
    const source = `
      <React.Fragment>
        <Component className="Inner" />
        <Icons className="Inner" />
        <Navigation className="Inner" />
      </React.Fragment>
    `

    expect(transformWithObjects(source)).toMatchSnapshot()
  })

  it('leaves previous class names but removes used one', () => {
    const source = `
      <React.Fragment>
        <Component className="Inner second" />
        <Icons className="Inner second" />
        <Navigation className="Inner second" />
      </React.Fragment>
    `

    expect(transformWithObjects(source)).toMatchSnapshot()
  })
})

describe('with option "attribute"', () => {
  const transformWithAttribute = source => transform(source, {
    attribute: 'styleName',
  })

  it('transforms only specified attributes', () => {
    const source = `
      <React.Fragment>
        <A className="Inner" styleName="Inner" />
        <B styleName="Inner" />
        <C className="Inner" />
      </React.Fragment>
    `

    expect(transformWithAttribute(source)).toMatchSnapshot()
  })
})
