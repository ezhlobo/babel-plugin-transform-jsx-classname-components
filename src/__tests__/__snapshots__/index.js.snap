// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not do anything with the simple code 1`] = `"<div><Component className=\\"hello\\" /></div>;"`;

exports[`transforms component if capitalized class name goes first 1`] = `
"<React.Fragment>
  <Component.Inner data-text=\\"extra\\" className=\\"extra\\" />
  <Component data-text=\\"extra\\" className=\\"extra Inner\\" />
</React.Fragment>;"
`;

exports[`transforms computed attributes 1`] = `
"<React.Fragment>
  <Component.Inner className={['simple'].join(' ')} />
  <Component.Inner className={[true && 'simple'].join(' ')} />
</React.Fragment>;"
`;

exports[`transforms with closing names 1`] = `
"<React.Fragment>
  <One>Hello</One>
  <Two.Inner>Hello</Two.Inner>
  <Three>Hello</Three>
  <Four.Inner><FourNested>Hello</FourNested></Four.Inner>
  <Five><FiveNested.Inner>Hello</FiveNested.Inner></Five>
  <Six.Inner><SixNested.Inner>Hello</SixNested.Inner></Six.Inner>
</React.Fragment>;"
`;

exports[`transforms with unusual names 1`] = `
"<React.Fragment>
  <A.B />
  <A0.B0 />
  <CamelCase.CamelCase />
  <CamelCase0.CamelCase0 />
  <CamelCase.$CamelCase />
  <Camel_Case._Snake_Case />
</React.Fragment>;"
`;

exports[`with option "objects" leaves previous class names but removes used one 1`] = `
"<React.Fragment>
  <Component className=\\"Inner second\\" />
  <Icons.Inner className=\\"second\\" />
  <Navigation.Inner className=\\"second\\" />
</React.Fragment>;"
`;

exports[`with option "objects" transforms only specified objects 1`] = `
"<React.Fragment>
  <Component className=\\"Inner\\" />
  <Icons.Inner />
  <Navigation.Inner />
</React.Fragment>;"
`;
